enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,586 @@
+   respectively of `infrastructure.config.openshift.io/v1` resource will result in adding 
+   labels and tags to all OpenShift managed GCP resources.
+
+### Risks and Mitigations
+- Risks in using google-beta provider:
+
+  Except for the additional features/functionality available for preview in the beta version,
+  google-beta provider is not different from the GA version google provider. google-beta
+  provider binary is embedded into installer as part of the installer build process and there
+  are no additional requirements made on the end user to use the beta version.
+
+  Below are some of the risks in using the google-beta provider.
+  1. google-beta is a preview version and doesn't always carry the SLAs or technical
+  support obligations.
+  2. google-beta provider always uses the beta API regardless of whether the
+  request contains any beta features.
+  3. google_compute_address API provides just the `labels` functionality in the beta version,
+  which is in beta since the initial [commit](https://github.com/hashicorp/terraform-provider-google-beta/commit/0fdc262183ed82e78824e42d05042a7b0aba7c8b)
+  in 2018, though the usual beta phase for a feature is six months.
+
+### Drawbacks
+- User-defined labels cannot be updated on an GCP resource which is not managed by an
+  operator. In this proposal, the changes proposed and developed will be part of
+  openshift-* namespace. External operators are not in scope.
+  User-defined labels can be updated on the following GCP resources.
+  - Compute Engine Instance
+  - Compute Address
+  - Compute Image
+  - Compute Forwarding Rule
+  - Storage Bucket
+  - DNS Zone
+
+- GCP Labels can be used as queryable annotations for resources, but can't be used
+  to set conditions on policies.
+
+- User-defined tags cannot be updated on a GCP resource which is not managed by an
+  operator. In this proposal, the changes proposed and developed will be part of
+  openshift-* namespace. External operators are not in scope.
+  User-defined tags can be updated on the following GCP resources.
+  - Compute Engine Instances
+  - Cloud Storage Buckets
+
+- Administrator will have to manually perform below label pertaining actions
+    1. removing the undesired labels/tags from the required resources.
+    2. update label/tags values of the required resources.
+    3. update labels/tags of the resources which are not managed by an operator.
+    4. update labels/tags of the resources for which update logic is not supported
+       by an operator.
+
+## Design Details
+
+### Open Questions
+- GCP supports 50 tags per resource and since OpenShift do not add any tags of its own,
+  users are allowed to define upto 50 tags. But the terraform and GO APIs for binding the
+  tags to the resources doesn't support batch operation and requires each tag to be added
+  individually to each resource which will increase the cluster creation time. What would
+  be the optimal number of tags a user can define?
```

Comment:
It looks like the google APIs limit the number of "requests" to 10 per second. If the tag additions cannot be batched we may need some timing operations to ensure that this quota isn't breached, or limit the number to 10.

Reply:
I was unaware of this, Thank you for sharing!
Yeah, I think limiting number of tags to 10 and timing the operations would be better. We did ask for a feedback from users, but no response yet. If more number of tags are required we could work on it while making this feature GA.

/cc @JoelSpeed @TrilokGeer

Reply:
Agree with @JoelSpeed , RPS should not be a limitation for total tags. It'd be helpful to note this limitation in product documentation. There might be an added installation delay when there are large number resources being configured at installation. For example, worker nodes being configured for installation (approximately 5s per resource instance). Would it affect any timeout configured for installer?

Reply:
@JoelSpeed I am still looking for rate limiting code in k8s/openshift. I also found a sub-repo of golang which provides has the package `golang.org/x/time/rate` implementing rate limit functionality, using which I was able to limit the number of requests made in a second.

Reply:
`Open Questions` section updated with the solution considered for TechPreview.

Reply:
Ok so we can rate limit on an individual component basis, but how does this work overall? If two operators are both rate limiting themselves, there is still a chance to hit the limit on the Google side right? I guess we can be best effort, but I'd be interested to know how the error comes back and if the operators are able to handle those well, any thoughts? Perhaps we don't need a rate limit ourselves if we can interpret the google errors sufficiently well

Reply:
Google errors does provide information on exceeding rate limit through http code 429 and the APIs support retry mechanism on particular http codes. I think it would be good to have both rate limiting mechanism of 8 requests/second and retry logic for http code 429, with initial backoff of 90seconds, since quota defined for Tag write operations is 600requests/minute and for maximum of 5minutes with retry period factor of 2. I did try out this after hitting the rate limit by adding 50 tags through 9 parallel routines and worked as expected, without rate limit exceeded errors.

Reply:
Ok cool, as long as we have a plan, that sounds good, can you make sure to incorporate those details into the enhancement

Reply:
Yeah, I have captured it [here](https://github.com/openshift/enhancements/pull/1217/files#diff-0c4f04b9b4a7dd08afbc3b7fcf8c4a15b3b4b31b51faaf43ecb33c0e9de4cdc4R353-R357)